tt_metal/impl/debug/dprint_server.cpp

```diff
@@ -206,25 +172,174 @@ static void print_tile_slice(ostream& stream, uint8_t* ptr, int hart_id) {
     // compute the buffer address for the requested hart
     uint32_t base_addr = PRINT_BUFFER_NC + hart_id*PRINT_BUFFER_SIZE;
 
     vector<uint32_t> initbuf = { DEBUG_PRINT_SERVER_STARTING_MAGIC };
     auto result = tt::llrt::read_hex_vec_from_core(chip_id, core, base_addr, 4);
     return (result[0] != initbuf[0]);
-} // check_init_magic_cleared
+} // CheckInitMagicCleared
+
+DebugPrintServerContext::DebugPrintServerContext() {
+    TT_ASSERT(inst == nullptr);
+    inst = this;
+
+    // Read hart mask + log file from rtoptions
+    uint32_t hart_mask = tt::llrt::OptionsG.get_dprint_riscv_mask();
+    string file_name = tt::llrt::OptionsG.get_dprint_file_name();
+
+    // Set the output stream according to RTOptions, either a file name or stdout if none specified.
+    if (file_name != "") {
+        outfile_ = new std::ofstream(file_name);
+    }
+    stream_ = outfile_ ? outfile_ : &cout;
+
+    stop_print_server_ = false;
+    mute_print_server_ = false;
+    new_data_last_iter_ = false;
+    server_killed_due_to_hang_ = false;
+
+    // Spin off the thread that runs the print server.
+    print_server_thread_ = new std::thread(
+        [this, hart_mask] { PollPrintData(hart_mask); }
+    );
+} // DebugPrintServerContext
+
+DebugPrintServerContext::~DebugPrintServerContext() {
+    // Signal the print server thread to finish
+    stop_print_server_ = true;
+
+    // Wait for the thread to end, with a timeout
+    auto future = std::async(std::launch::async, &std::thread::join, print_server_thread_);
+    if (future.wait_for(std::chrono::seconds(2)) == std::future_status::timeout) {
+        TT_FATAL(false && "Timed out waiting on debug print thread to terminate.");
+    }
+    delete print_server_thread_;
+    print_server_thread_ = nullptr;
+
+    if (outfile_) {
+        outfile_->close();
+        delete outfile_;
+    }
+    inst = nullptr;
+} // ~DebugPrintServerContext
+
+void DebugPrintServerContext::WaitForPrintsFinished() {
+    // Simply poll the flag every few ms to check whether new data is still being processed,
+    // or whether any cores are waiting for a signal to be raised.
+    // TODO(dma): once we have access to the device is there a way we can poll the device to
+    // check whether more print data is coming?
+    do {
+        // No need to await if the server was killed already due to a hang.
+        if (server_killed_due_to_hang_)
+            break;
+        std::this_thread::sleep_for(std::chrono::milliseconds(5));
+    } while (hart_waiting_on_signal_.size() > 0 || new_data_last_iter_);
+} // WaitForPrintsFinished
+
+void DebugPrintServerContext::AttachDevice(Device* device) {
+    chip_id_t device_id = device->id();
+
+    // A set of all valid worker cores, used for checking the user input. Note that the core coords
+    // here are physical, which matches the user input for configuring the print server.
+    auto compare_coords = [](const CoreCoord& a, const CoreCoord& b){
```

Comment:
Doesn't CoreCoord already have these defined? it's in `tt_xy_pair.h`

Reply:
Missed this yeah, removed this redundant function in latest commit